src/services/pasteEdits.ts

```diff
@@ -0,0 +1,123 @@
+    cancellationToken: CancellationToken,
+    changes: textChanges.ChangeTracker,
+) {
+    const statements: Statement[] = [];
+
+    let start = 0;
+    let newText = "";
+    pasteLocations.forEach((location, i) => {
+        if (i === pasteLocations.length - 1) {
+            newText += targetFile.text.slice(start, location.pos) + pastedText[i] + targetFile.text.slice(location.end);
+        }
+        else {
+            newText += targetFile.text.slice(start, location.pos) + pastedText[i];
+            start = location.end;
+        }
+    });
+    host.runWithTemporaryFileUpdate?.(targetFile.fileName, newText, (updatedProgram, originalProgram, updatedFile) => {
+        const importAdder = codefix.createImportAdder(updatedFile, updatedProgram, preferences, host);
+        if (copiedFrom?.range) {
+            Debug.assert(copiedFrom.range.length === pastedText.length);
+            copiedFrom.range.forEach(copy => {
+                addRange(statements, copiedFrom.file.statements, getLineOfLocalPosition(copiedFrom.file, copy.pos), getLineOfLocalPosition(copiedFrom.file, copy.end) + 1);
+            });
+            const usage = getUsageInfo(copiedFrom.file, statements, originalProgram!.getTypeChecker(), getExistingLocals(updatedFile, statements, originalProgram!.getTypeChecker()));
+            Debug.assertIsDefined(originalProgram);
+            const useEsModuleSyntax = !fileShouldUseJavaScriptRequire(targetFile.fileName, originalProgram, host, !!copiedFrom.file.commonJsModuleIndicator);
+            addExportsInOldFile(copiedFrom.file, usage.targetFileImportsFromOldFile, changes, useEsModuleSyntax);
+            addTargetFileImports(copiedFrom.file, usage.oldImportsNeededByTargetFile, usage.targetFileImportsFromOldFile, originalProgram.getTypeChecker(), updatedProgram, importAdder);
+            importAdder.writeFixes(changes, getQuotePreference(copiedFrom.file, preferences));
+        }
+        else {
+            const context: CodeFixContextBase = {
+                sourceFile: updatedFile,
+                program: originalProgram!,
+                cancellationToken,
+                host,
+                preferences,
+                formatContext,
+            };
+            forEachChild(updatedFile, function cb(node) {
+                if (isIdentifier(node)) {
+                    if (!originalProgram?.getTypeChecker().resolveName(node.text, node, SymbolFlags.All, /*excludeGlobals*/ false)) {
+                        // generate imports
+                        importAdder.addImportForUnresolvedIdentifier(context, node, /*useAutoImportProvider*/ true);
+                    }
+                }
+                else {
+                    node.forEachChild(cb);
+                }
+            });
+            importAdder.writeFixes(changes, getQuotePreference(targetFile, preferences));
+        }
+    });
+
+    if (pastedText.length !== 1) {
+        Debug.assert(pastedText.length === pasteLocations.length);
+    }

```

Comment:
Maybe this should happen at the beginning? VS Code can handle this scenario, though I think it's fine if we don't. Will the extension ensure it doesn't ask us for paste edits when this is true?

Reply:
The last time I checked with Matt, he mentioned that he wanted these edits along with the import edits that we provide, but I can remove if needed.
As for this being true, I believe that this would be checked on the vscode side, @mjbvz ?

Reply:
Yeah, you are supposed to apply these edits; I remember Matt clarifying that. I meant that VS Code has no problem pasting two copy locations (`pastedText.length === 2`) into three paste locations (`pasteLocations.length === 3`). What happens is that each `pasteLocation` gets `pastedText.join("\n")` inserted, but we're throwing.